### DISCUSSION TITLE
Typescript Rewrite

### DISCUSSION THREAD

**Discussion by adambrangenberg** [PrismarineJS/mineflayer#2964] [Ideas]
Hey, I opened this discussion because:
1) I'm not sure in what order to rewrite things to get the types of the basis while creating ones for higher packages?
2) If the project even wants to use Typescript?
 
The benefits of using Typescript would be a higher code quality because with static typing many bugs are being catched while developing and easier development because of improved editor support and documentation.

**Comment by rom1504** [comment]
Hi,

Typescript has a high entry cost (compilation making everything slower and
making usage of dependencies harder) and low benefits.

If you manage to make the code much higher quality by rewriting a
significant package to it, it might be convincing.

Otherwise we're staying in JavaScript.

**Reply by adambrangenberg** [reply]
Okay, can you recommend any module to do that?
Edit: There are a few tools that compile pretty fast, like swc or the regular typescript compiler. So I'm not sure what you mean by that?

**Reply by rom1504** [reply]
Probably mineflayer

**Reply by adambrangenberg** [reply]
Hey, I've ran into a few issues like some dependencies not having used methods implemented (e.g. anvil.js using .close from prismarine-windows, which doesn't seem to have such method). Do you know anything about why that is? It's hard to implement types for things that seem to be non-existent :/

**Comment by AugustinMauroy** [comment]
If we rewrite it we should use ESM import/export

**Comment by AugustinMauroy** [comment]
@rom1504 Create an branch name `major/typescript`. To allow the contributor to code.

**Reply by adambrangenberg** [reply]
That would be great. I already did the anvil module, but still waiting for the typing in prismarine-item to get merged.

**Reply by AugustinMauroy** [reply]
WDYM by typing? migrate to typescript or use just type ?

**Reply by adambrangenberg** [reply]
Exporting some things, etc: https://github.com/PrismarineJS/prismarine-item/pull/87
